### PR TITLE
Convert SQLAlchemySchemaInfo to a dataclass

### PR DESCRIPTION
**Pull request by chewselene**
In addition to converting the namedtuple to a dataclass, this PR also:

-  adds a comment about the difference between `SQLAlchemySchemaInfo` and `SQLSchemaInfo`
-  updates a test file to make it mypy compliant

This is a part of the refactor in #1008

### CODE REVIEW

File: graphql_compiler/schema/schema_info.py [hunk 1]
```diff
@@ -226,56 +226,62 @@ def _create_sql_schema_info(
...
+
+    It describes the tables that correspond to each type (object type or interface type),
+    and gives instructions on how to perform joins for each vertex field. The property fields on
+    each type are implicitly mapped to columns with the same name on the corresponding table.
+
+    Notes:
+    - RootSchemaQuery is a special type that does not need a corresponding table.
+    - Builtin types like __Schema, __Type, etc. don't need corresponding tables.
+    - Builtin fields like _x_count do not need corresponding columns.
+    TODO: This class is essentially the same as SQLSchemaInfo. SQLSchemaInfo is part of an
```
Comment (bojanserafimov) [inline graphql_compiler/schema/schema_info.py:241]:
```suggestion

    TODO: This class is essentially the same as SQLSchemaInfo. SQLSchemaInfo is part of an
```

**Review by bojanserafimov — DISMISSED** [review]
Looks good, one nit.

File: graphql_compiler/compiler/emit_sql.py [hunk 2]
```diff
@@ -632,7 +632,7 @@ def _get_fold_outputs(self) -> List[Label]:
 
     def add_traversal(
         self,
-        join_descriptor: DirectJoinDescriptor,
+        join_descriptor: JoinDescriptor,
```
Comment (chewselene) [inline graphql_compiler/compiler/emit_sql.py:635]: I thought this seemed more reasonable than adding a type check in the test that was failing, but I'm not very familiar with the composite edge stuff so @bojanserafimov let me know if this doesn't actually make sense!

Comment (bojanserafimov) [reply graphql_compiler/compiler/emit_sql.py:635]: This is not a user-friendly message (which edge uses composite keys?) so we should either make it user friendly, and remove the similar error check in `CompilationState.traverse`, or turn it into an assertion error.

Comment (bojanserafimov) [reply graphql_compiler/compiler/emit_sql.py:635]: Alternatively, just add a comment at the other message explaining why its necessary even though the internal function does the same check: that the external check returns a friendlier message

Comment (bojanserafimov) [reply graphql_compiler/compiler/emit_sql.py:635]: Ideally we'd have all our errors return a text span in the original query pointing at where the error is (so the UI can underline in red), but that's some polish work I don't think we'll get to

**Review by bojanserafimov — APPROVED** [review]

**Review by Aytac998 — COMMENTED** [review]
A
